packages/cardano-services/src/Asset/CardanoTokenRegistry.ts

```diff
@@ -0,0 +1,171 @@
+import { Asset, Cardano } from '@cardano-sdk/core';
+import { InMemoryCache } from '../InMemoryCache';
+import { Logger } from 'ts-log';
+import { TokenMetadataService } from './types';
+import axios, { AxiosInstance } from 'axios';
+
+const DEFAULT_METADATA_SERVER_URI = 'https://tokens.cardano.org';
+
+interface NumberValue {
+  value?: number;
+}
+
+interface StringValue {
+  value?: string;
+}
+
+interface TokenMetadataServiceRecord {
+  decimals?: NumberValue;
+  description?: StringValue;
+  logo?: StringValue;
+  name?: StringValue;
+  subject?: string;
+  ticker?: StringValue;
+  url?: StringValue;
+}
+
+/**
+ * Configuration options for CardanoTokenRegistry
+ */
+export interface CardanoTokenRegistryConfiguration {
+  /**
+   * The cache TTL in seconds. Default: 10 minutes.
+   */
+  cacheTTL?: number;
+
+  /**
+   * The Cardano Token Registry public API base URL. Default: https://tokens.cardano.org
+   */
+  metadataServerUri?: string;
+}
+
+interface CardanoTokenRegistryConfigurationWithRequired extends CardanoTokenRegistryConfiguration {
+  cacheTTL: number;
+  metadataServerUri: string;
+}
+
+/**
+ * Dependencies that are need to create CardanoTokenRegistry
+ */
+export interface CardanoTokenRegistryDependencies {
+  /**
+   * The cache engine. Default: InMemoryCache with CardanoTokenRegistryConfiguration.cacheTTL as default TTL
+   */
+  cache?: InMemoryCache;
+
+  /**
+   * The logger object
+   */
+  logger: Logger;
+}
+
+/**
+ * TokenMetadataService implementation using Cardano Token Registry public API
+ */
+export class CardanoTokenRegistry implements TokenMetadataService {
+  /**
+   * The axios client used to retrieve metadata from API
+   */
+  #axiosClient: AxiosInstance;
+
+  /**
+   * The in memory cache engine
+   */
+  #cache: InMemoryCache;
+
+  /**
+   * The logger object
+   */
+  #logger: Logger;
+
+  constructor({ cache, logger }: CardanoTokenRegistryDependencies, config: CardanoTokenRegistryConfiguration = {}) {
+    const defaultConfig: CardanoTokenRegistryConfigurationWithRequired = {
+      cacheTTL: 600,
+      metadataServerUri: DEFAULT_METADATA_SERVER_URI,
+      ...config
+    };
+
+    this.#cache = cache || new InMemoryCache(defaultConfig.cacheTTL);
+    this.#axiosClient = axios.create({ baseURL: defaultConfig.metadataServerUri });
+    this.#logger = logger;
+  }
+
+  shutdown() {
+    this.#cache.shutdown();
```

Comment:
Yes, the management of the dependency should be limited to when the current context created a state change, ie. a class instance started it so it's now responsible to stop it.

Reply:
Right. In this specific case it would be good to have an explicit `InMemoryCache.initialize` or `.start`, so that it's more obvious on where to shut it down.

Reply:
As I told in another thread @rhyslbw and @mkazlauskas , I don't like to force the caller to perform _not required calls_ as `InMemoryCache.initialize()` and `InMemoryCache.start()` would be going in that direction.
Anyway if that's the requirement for our standard, this reason is much more stronger than anyone else bound to what I like or not.

Before proceeding, however, we have an open discussion about the definition of `Provider` and another one about the right placement of `RunningModule`... I think it's better first of all to consolidate the standard, then we could refactor both `InMemoryCache` and `CardanoTokenRegistry` classes.